plugins/com.google.cloud.tools.eclipse.appengine.compat.cte13/src/com/google/cloud/tools/eclipse/appengine/compat/cte13/CloudToolsEclipseProjectNotifier.java

```diff
@@ -0,0 +1,151 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.tools.eclipse.appengine.compat.cte13;
+
+import com.google.cloud.tools.eclipse.util.status.StatusUtil;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Ordering;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.logging.Logger;
+import org.eclipse.core.resources.IProject;
+import org.eclipse.core.resources.IWorkspace;
+import org.eclipse.core.resources.ResourcesPlugin;
+import org.eclipse.core.resources.WorkspaceJob;
+import org.eclipse.core.runtime.CoreException;
+import org.eclipse.core.runtime.IProgressMonitor;
+import org.eclipse.core.runtime.IStatus;
+import org.eclipse.core.runtime.MultiStatus;
+import org.eclipse.core.runtime.Status;
+import org.eclipse.core.runtime.SubMonitor;
+import org.eclipse.core.runtime.jobs.Job;
+import org.eclipse.jface.dialogs.MessageDialog;
+import org.eclipse.swt.widgets.Shell;
+import org.eclipse.ui.IStartup;
+import org.eclipse.ui.IWorkbench;
+import org.eclipse.ui.PlatformUI;
+
+/**
+ * Looks for projects using the per-library classpath container approach used in versions prior
+ * Cloud Tools for Eclipse 1.4, and notifies the user to upgrade.
+ */
+public class CloudToolsEclipseProjectNotifier implements IStartup {
+  private static final Logger logger =
+      Logger.getLogger(CloudToolsEclipseProjectNotifier.class.getName());
+  private IWorkbench workbench;
+  private IWorkspace workspace;
+
+  @Override
+  public void earlyStartup() {
+    workbench = PlatformUI.getWorkbench();
+    workspace = ResourcesPlugin.getWorkspace();
+
+    Job projectUpdater = new WorkspaceJob("Updating projects for Cloud Tools for Eclipse") {
+      @Override
+      public IStatus runInWorkspace(IProgressMonitor monitor) throws CoreException {
+        SubMonitor progress = SubMonitor.convert(monitor, 40);
+        progress.subTask("Searching for projects requiring updating");
+        Collection<IProject> projects = findCandidates(progress.newChild(10));
+        if (projects.isEmpty()) {
+          return Status.OK_STATUS;
+        }
+        projects = promptUser(projects, progress.newChild(5));
+        if (projects.isEmpty()) {
+          return Status.OK_STATUS;
+        }
+        progress.subTask("Updating projects");
+        return upgradeProjects(projects, progress.newChild(25));
+      }
+    };
+    projectUpdater.setRule(workspace.getRoot());
+    projectUpdater.setUser(true);
+    projectUpdater.schedule(500);
+  }
+
+  /**
+   * Scan the current projects to identify those requiring upgrading.
+   */
+  protected Collection<IProject> findCandidates(SubMonitor progress) {
```

Comment:
All the methods except the one above are `protected`. Intended?

Reply:
Nope